### PR TITLE
feat: Portkey integration with Llama_index

### PR DESCRIPTION
**Pull request by noble-varghese**
# Description

This pull request introduces the Portkey integration to the Llama_index repository, enhancing its production capabilities. Portkey offers a comprehensive set of tools and features to optimize the performance and reliability of AI applications.

### Key Features of Portkey's Integration with Llamaindex:
1. **AI Gateway**:
    - **Automated Fallbacks & Retries**: Ensure your application remains functional even if a primary service fails.
    - **Load Balancing**: Efficiently distribute incoming requests among multiple models.
    - **Semantic Caching**: Reduce costs and latency by intelligently caching results.
2. **Observability**:
    - **Logging**: Keep track of all requests for monitoring and debugging.
    - **Requests Tracing**: Understand the journey of each request for optimization.
    - **Custom Tags**: Segment and categorize requests for better insights.

### Changes Made:
Incorporated a new LLM (Large Language Model) within the LLMs folder, which inherits from the existing CustomLLM class. This new LLM is responsible for integrating and utilizing Portkey's enhanced production capabilities while maintaining consistency with the repository's coding standards.

### Dependencies:
This integration introduces one external library called "Rubeus," which is added to the list of dependencies. Rubeus is an integral part of Portkey and contributes to its internal functioning.

Fixes # (issue)

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: llama_index/llms/portkey.py [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+    LLMBase,
+    RubeusModes,
+    RubeusModesLiteral,
+    ProviderTypes,
+    ProviderTypesLiteral,
+    RubeusCacheType,
+    RubeusCacheLiteral,
+    Message,
+    RubeusResponse,
+)
```
Comment (logan-markewich) [inline llama_index/llms/portkey.py:17]: If you want to keep these imports for typing, take a look at the marvin extractor for an example as well
https://github.com/jerryjliu/llama_index/blob/main/llama_index/node_parser/extractors/marvin_metadata_extractor.py

**Review by logan-markewich — COMMENTED** [review]
Awesome, looks good so far! The main concern is I'd rather not add anything to the core dependencies. Instead, a try/catch with an import error in the init is the preferred approach

File: llama_index/llms/portkey.py [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+    get_llm,
+)
+from llama_index.llms.generic_utils import (
+    completion_to_chat_decorator,
+    chat_to_completion_decorator,
+    stream_completion_to_chat_decorator,
+    stream_chat_to_completion_decorator,
+)
+
+try:
```
[the remark this answers is not in the document]
Comment (noble-varghese) [reply llama_index/llms/portkey.py:41]: replaced this try/except block with the import from `llama_index.bridge.pydantic`

File: llama_index/llms/portkey.py [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+    provider: Optional[Union[ProviderTypes, ProviderTypesLiteral]] = Field(
+        default=ProviderTypes.OPENAI
+    )
+    llm: LLMBase = Field(description="LLM parameter", default_factory=dict)
+
+    llms: List[LLMBase] = Field(description="LLM parameters", default_factory=list)
+
+    _client: Rubeus = PrivateAttr()
+    _portkey_response: Any = PrivateAttr()
+    _model: Any = PrivateAttr()
```
[the remark this answers is not in the document]
Comment (noble-varghese) [reply llama_index/llms/portkey.py:70]: This is not used. I have removed the un-used class attributes.

File: llama_index/llms/portkey.py (hunk 1 above)
Comment (noble-varghese) [reply llama_index/llms/portkey.py:17]: Made the changes by moving them under the TYPE_CHECKING condition.

**Comment by noble-varghese** [comment]
Updated the code with the required dependency import errors. 
This seems to fail the workflows - Build Package and Unit Testing. I hope that's expected. 
The error is the import failure in both of those cases.

File: llama_index/llms/portkey.py [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+
+
+class Portkey(CustomLLM):
+    """_summary_
+
+    Args:
+        LLM (_type_): _description_
+    """
+
+    try:
```
Comment (logan-markewich) [inline llama_index/llms/portkey.py:50]: Sorry if it was confusing -- I think this should go into `__init__`

Comment (logan-markewich) [reply llama_index/llms/portkey.py:50]: Then, build/unit tests should pass

Comment (noble-varghese) [reply llama_index/llms/portkey.py]: Yes. That fixed it. Thanks :)

**Comment by logan-markewich** [comment]
Thanks again for the work on this @noble-varghese ! Going to merge now 🙌

**Comment by noble-varghese** [comment]
Pleasure has been mine @logan-markewich . Looking forward to contribute more to Llama_index. You guys are doing an amazing job 🙌

**Pull request merged**